### PR TITLE
Fix NetFramingBase based transports Abort code path

### LINKED ISSUE #1345: [Bug]: NetTCP connection doesn't send RST packet when timeout
### Duplicate ?

- [X] I have searched issues/discussions and did not find other issues/discussions reporting this bug.

### Product version

1.5.1

### Describe expected behavior

When server (service) timeouts (disconnects / closes connection) then RST packet should be sent to the client so it knows about being disconnected.

### Describe actual behavior

When server timeouts then no RST packet is sent so client thinks it is still connected. but if it tries to get any data from the server, it timeouts (after client timeout). 

I have tried to create server and client in NET6 using CoreWCF and also in Framework 4.7.2 to compare behaviour. 

**Framework 4.7.2 behaviour:**

- Server starts, client connects - no problem 
- I set timeout of server to 10 seconds. Timeout of client to 5 seconds. 
- If I do nothing for 10 seconds, server timeouts and informs client about it
    - Client disconnects
    - It is visible in Wireshark

image

Events on server side:

image

**NET6 behaviour:**
- Server starts, client connects - no problem
- Timeouts are again - 10 seconds at server side, 5 seconds at client side
- If I do nothing for 10 seconds, I see that server closes connection, but nothing is sent to the client and therefore connection looks active, even though, it is not. 
    - If client tries to call something from service, it doesn't reach it because connection is off and timeouts in 5 seconds as not reachable. 

Server side events:
image

But nothing is sent to the client and therefore connection looks still on:

image

But, if I try to call method from client again, it doesn't reach it. And timeouts in 5 seconds (client timeout)

image
image

We can see in the server logs that it didn't reach it at all so connection was closed. Client just didn't know:
image

### Which binding

NetTcp

### security

Transport

### Which .NET version

.NET 6

### Which os platform

Windows

### Code snippet used to reproduce the issue

```c#
// WCF Service

[ServiceContract]
public interface ITestWcfService
{
    [OperationContract]
    string GetData(int value);
}

[ServiceBehavior(InstanceContextMode = InstanceContextMode.PerSession, ConcurrencyMode = ConcurrencyMode.Multiple)]
public class TestWcfService : ITestWcfService
{
    public TestWcfService() //only instantiated by WCF
    {
        Console.WriteLine("Constructor - NET6");
        OperationContext.Current.InstanceContext.Faulted += Channel_Faulted;
        OperationContext.Current.InstanceContext.Closed += Channel_Closed;
        OperationContext.Current.InstanceContext.Opened += InstanceContext_Opened;
    }

    public string GetData(int value)
    {
        Console.WriteLine("GetData - NET6");
        try
        {
            return $"{value}";
        }
        catch (Exception ex)
        {
            throw;
        }
    }

    private void InstanceContext_Opened(object sender, EventArgs e)
    {
        Console.WriteLine("Opened - NET6");
    }

    private void Channel_Faulted(object sender, EventArgs e)
    {
        Console.WriteLine("Faulted - NET6");
    }

    private void Channel_Closed(object sender, EventArgs e)
    {
        Console.WriteLine("Closed - NET6");
    }
}

// Server:

private static WebApplication _app;
public Form1()
{
    InitializeComponent();

    var builder = WebApplication.CreateBuilder();
    builder.WebHost.UseNetTcp(11999);
    builder.Services.AddServiceModelServices();
    builder.Services.AddServiceModelMetadata();

    _app = builder.Build();

    var binding = new NetTcpBinding(SecurityMode.Transport)
    {
        CloseTimeout = TimeSpan.FromSeconds(10),
        OpenTimeout = TimeSpan.FromSeconds(10),
        ReceiveTimeout = new TimeSpan(0, 0, 0, 10),
        SendTimeout = TimeSpan.FromSeconds(10)
    };
    
    _app.UseServiceModel(serviceBuilder =>
    {
        serviceBuilder.AddService<TestWcfService>(options =>
        {
            options.DebugBehavior.IncludeExceptionDetailInFaults = true;
        });

        serviceBuilder.AddServiceEndpoint<TestWcfService, ITestWcfService>(binding, "net.tcp://localhost:11999/TestWcfService2");
    });

    _app.StartAsync().Wait();
}

// Client:

private ITestWcfService client;
public Form1()
{
    InitializeComponent();

    IClientChannel channel = null;
    var binding = new NetTcpBinding()
    {
        CloseTimeout = TimeSpan.FromSeconds(5),
        OpenTimeout = TimeSpan.FromSeconds(5),
        ReceiveTimeout = new TimeSpan(0, 0, 0, 5),
        SendTimeout = TimeSpan.FromSeconds(5),
    };

    var factory = new ChannelFactory<ITestWcfService>(binding, new EndpointAddress("net.tcp://localhost:11999/TestWcfService2"));
    factory.Open();
    try
    {
        client = factory.CreateChannel();
    }
    finally
    {
    }
}

private void button1_Click(object sender, EventArgs e)
{
    client.GetData(4);
}
```

### Stacktrace if any

_No response_

### PR DESCRIPTION
**Pull request by mconnew**
Fixes #1345

There's a subtlety with SocketConnection in asp.net core that if you call ConnectionContext.Abort() without an exception, it cleanly closes the connection using Socket.Shutdown, but if you provide an exception, it does actually abort the socket.
This change also provides a dummy exception when calling Abort without an exception so that abort semantics are always followed (RST packet sent). This makes other existing Abort calls which don't have an exception actually reset the connection instead of closing it.

### CODE REVIEW

File: src/CoreWCF.Http/tests/ThreadPoolConfigInitializer.cs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+// The .NET Foundation licenses this file to you under the MIT license.
+
+#if NET6_0_OR_GREATER
+using System;
+using System.Runtime.CompilerServices;
+using System.Threading;
+
+namespace CoreWCF.Http.Tests
+{
+    internal static class ThreadPoolConfigInitializer
```
Comment (g7ed6e) [inline src/CoreWCF.Http/tests/ThreadPoolConfigInitializer.cs:11]: Is this module needed ? It looks like it has been used to debug things.

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/NetTcpOptions.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+
+using System.Collections.Generic;
+using System;
+
+namespace CoreWCF.Channels
+{
+    public class NetTcpOptions
+    {
+        internal List<TcpListenOptions> CodeBackedListenOptions { get; } = new List<TcpListenOptions>();
+        public IServiceProvider ApplicationServices { get; internal set; }
```
Comment (g7ed6e) [inline src/CoreWCF.NetTcp/src/CoreWCF/Channels/NetTcpOptions.cs:12]: Does this getter require to be public ?

File: src/testEnvironments.json [hunk 3]
```diff
@@ -0,0 +1,25 @@
+{
```
Comment (g7ed6e) [inline src/testEnvironments.json:1]: Do we need 2 copies of this file ? (there is one copy in the root folder and one in the src folder)

**Review by g7ed6e — APPROVED** [review]

**Comment by mconnew** [comment]
I've got one more change to push before I merge, and that's to drop the new nuget feed (xunit dev feed) and reference the latest xunit nuget packages once they are published.  

To give a summary of what else is in this PR besides they actual product fix, I added the ability to run unit tests on WSL2 or docker images to validate/easily debug tests running on Linux from Visual Studio. That required updating to the latest xunit, which broke things, which resulted in a very productive back and forth with the xunit project to introduce a new parallelism mode which should get rid of flakey test timeouts from here on.

File: src/testEnvironments.json (hunk 3 above)
Comment (mconnew) [reply src/testEnvironments.json]: Fixed that. I originally tried it in the src folder, but the constraints of the remote test execution are that the built binaries must be in a subfolder of the folder that the solution file lives in. This is why I moved the .sln files and the testEnvironments.json file to the root of the repo.

File: src/Common/tests/XunitTestAssemblyAttribute.cs [hunk 4]
```diff
@@ -2,4 +2,4 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 [assembly: Xunit.CollectionBehavior(MaxParallelThreads = -1)]
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/Common/tests/XunitTestAssemblyAttribute.cs]: I removed the attribute method of configuring this and switched to xunit runnner json files.

File: src/Common/UnitTests.Common/Xunit/ConcurrencyLimitingTestAssemblyRunner.cs [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+            _semaphore = new SemaphoreSlim(Environment.ProcessorCount);
+        }
+
+        protected override async Task<RunSummary> RunTestCollectionAsync(IMessageBus messageBus, ITestCollection testCollection, IEnumerable<IXunitTestCase> testCases, CancellationTokenSource cancellationTokenSource)
+        {
+            await _semaphore.WaitAsync(cancellationTokenSource.Token);
+            try
+            {
+                return await base.RunTestCollectionAsync(messageBus, testCollection, testCases, cancellationTokenSource).ConfigureAwait(false);
+                //return await new SyncContextTestCollectionRunner(testCollection, testCases, DiagnosticMessageSink, messageBus, TestCaseOrderer, new ExceptionAggregator(Aggregator), cancellationTokenSource).RunAsync();
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/Common/UnitTests.Common/Xunit/ConcurrencyLimitingTestAssemblyRunner.cs:28]: I thought I had deleted all these Xunit extensibility implementations. They'll be gone in my next iteration.

File: src/CoreWCF.NetTcp/src/CoreWCF/Channels/NetTcpOptions.cs (hunk 2 above)
Comment (mconnew) [reply src/CoreWCF.NetTcp/src/CoreWCF/Channels/NetTcpOptions.cs:12]: Yes, it mirrors how Kestrel handles this and enables you to pull things from DI in your delegate. E.g. you can do this:
```c#
webHost.UseNetTcp((NetTcpOptions options) =>
{
    IConfiguration configuration = options.ApplicationServices.GetRequiredService<IConfiguration>();
    // pull base address from config
    options.Listen(baseAddressFromConfig);
});
```

**Comment by mconnew** [comment]
@g7ed6e, can you take one more look, this should be ready to merge now. I believe I answered all your questions.

**Review by g7ed6e — APPROVED** [review]
There is one MS Build property that is defined twice with the same value but different conditions which may be aggregated to only one. Despite that LGTM.

File: src/CoreWCF.Http/tests/Helpers/ServiceHelper.cs [hunk 6]
```diff
@@ -180,72 +165,39 @@ public static Binding GetBufferedModHttp1Binding()
...
-                    options.Authentication.Schemes = Microsoft.AspNetCore.Server.HttpSys.AuthenticationSchemes.None;
-                    options.MaxConnections = null;
-                    options.MaxRequestBodySize = 30000000;
-                    options.UrlPrefixes.Add("http://localhost:8085");
-                })
-                .UseStartup<TStartup>();
+            CreateBaseWebHostBuilder<TStartup>(outputHelper, callerMethodName)
+            .UseHttpSys(options =>
+            {
+                options.AllowSynchronousIO = true;
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/CoreWCF.Http/tests/Helpers/ServiceHelper.cs:191]: Only for streamed transfer mode due to the serializers only being synchronous and directly reading from the stream.

File: src/CoreWCF.Http/tests/Helpers/ServiceHelper.cs [hunk 7]
```diff
@@ -261,8 +213,38 @@ public static Binding GetBufferedModHttp1Binding()
...
+#if DEBUG
+                    if (outputHelper != default)
+                        logging.AddProvider(new XunitLoggerProvider(outputHelper, callerMethodName));
+                    logging.AddFilter("Default", LogLevel.Debug);
+                    logging.AddFilter("Microsoft", LogLevel.Debug);
+                    logging.SetMinimumLevel(LogLevel.Debug);
+                    logging.AddDebug();
+#endif // DEBUG
+                })
+                .UseDefaultServiceProvider((context, options) =>
```
[the remark this answers is not in the document]
Comment (mconnew) [reply src/CoreWCF.Http/tests/Helpers/ServiceHelper.cs:242]: I just copied this from the code which creates the default web host builder.

File: Directory.Build.props [hunk 8]
```diff
@@ -38,12 +43,12 @@
...
   <PropertyGroup>
     <RunUnitTestsConcurrently Condition="$(IsTestProject)">true</RunUnitTestsConcurrently>
-  </PropertyGroup>  
+  </PropertyGroup>
   <PropertyGroup >
     <RunUnitTestsConcurrently Condition="$(IsTestProject) AND ($(IsMetadataProject) OR $(IsMsmqProject) OR $(IsQueueProject) OR $(IsKafkaProject))">false</RunUnitTestsConcurrently>
   </PropertyGroup>
   <PropertyGroup >
-    <RunUnitTestsConcurrently Condition="$(IsTestProject) AND $(IsHttpProject)">ExceptNetFx</RunUnitTestsConcurrently>
+    <RunUnitTestsConcurrently Condition="$(IsTestProject) AND $(IsHttpProject) AND $(TargetFramework.StartsWith('net4'))">false</RunUnitTestsConcurrently>
```
[the remark this answers is not in the document]
Comment (mconnew) [reply Directory.Build.props:51]: Good idea. I kept them separate initially to make it easier to switch back and forth between the old and new ways. I'll make that change then merge.

**Pull request merged**